util/TLSProxy/Message.pm

```diff
@@ -198,14 +198,14 @@ sub get_messages
         print "  [".$record->decrypt_data."]\n";
     } elsif ($record->content_type == TLSProxy::Record::RT_ALERT) {
         my ($alertlev, $alertdesc) = unpack('CC', $record->decrypt_data);
-        #All alerts end the test
-        $end = 1;
         #A CloseNotify from the client indicates we have finished successfully
         #(we assume)
-        if (!$server && $alertlev == AL_LEVEL_WARN
+        if (!$end && !$server && $alertlev == AL_LEVEL_WARN
             && $alertdesc == AL_DESC_CLOSE_NOTIFY) {
             $success = 1;
         }
+        #All alerts end the test
+        $end = 1;
```

Comment:
Can you wait for a mutual close_notify?

In the long run, I want to switch the proxy to use test code (https://github.com/openssl/openssl/blob/master/test/handshake_helper.c), and not the command-line apps, which should make all this easier to control.